hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -0,0 +1,208 @@
+type StorageTypeFactory struct {
+	service                    string                                                  // Name of the group we're handling (used mostly for logging)
+	types                      astmodel.Types                                          // All the types for this group
+	pendingStorageConversion   astmodel.TypeNameQueue                                  // Queue of types that need storage variants created for them
+	pendingConversionInjection astmodel.TypeNameQueue                                  // Queue of types that need conversion functions injected
+	pendingMarkAsHubVersion    astmodel.TypeNameQueue                                  // Queue of types that need to be flagged as the hub storage version
+	idFactory                  astmodel.IdentifierFactory                              // Factory for creating identifiers
+	typeConverter              *TypeConverter                                          // a utility type type visitor used to create storage variants
+	functionInjector           *FunctionInjector                                       // a utility used to inject functions into definitions
+	resourceHubMarker          *HubVersionMarker                                       // a utility used to mark resources as Storage Versions
+	conversionMap              map[astmodel.PackageReference]astmodel.PackageReference // Map of conversion links for creating our conversion graph
+}
+
+// NewStorageTypeFactory creates a new instance of StorageTypeFactory ready for use
+func NewStorageTypeFactory(service string, idFactory astmodel.IdentifierFactory) *StorageTypeFactory {
+	result := &StorageTypeFactory{
+		service:                    service,
+		types:                      make(astmodel.Types),
+		pendingStorageConversion:   astmodel.MakeTypeNameQueue(),
+		pendingConversionInjection: astmodel.MakeTypeNameQueue(),
+		pendingMarkAsHubVersion:    astmodel.MakeTypeNameQueue(),
+		idFactory:                  idFactory,
+		conversionMap:              make(map[astmodel.PackageReference]astmodel.PackageReference),
+		functionInjector:           NewFunctionInjector(),
+		resourceHubMarker:          NewHubVersionMarker(),
+	}
+
+	result.typeConverter = NewTypeConverter(result.types)
+
+	return result
+}
+
+// Add the supplied type definition to this factory
+func (f *StorageTypeFactory) Add(def astmodel.TypeDefinition) {
+	f.types.Add(def)
+
+	// Add to our queue of types requiring storage variants
+	f.pendingStorageConversion.Enqueue(def.Name())
+}
+
+// Types returns types contained by the factory, including all new storage variants and modified
+// api types. If any errors occur during processing, they're returned here.
+func (f *StorageTypeFactory) Types() (astmodel.Types, error) {
+	err := f.process()
+	if err != nil {
+		return nil, err
+	}
+
+	return f.types, nil
+}
+
+func (f *StorageTypeFactory) process() error {
+	err := f.pendingStorageConversion.Process(f.createStorageVariant)
```

Comment:
I'm not sure the queues are buying you much here - each one is only populated from one place, and they're processed one after the other, so they could have just been slices. That said, they don't really hurt either. I was just expecting the population logic to be more complicated.

Reply:
Agree: it wasn't clear to me if the queue type was actually needed here or not (I think not?)

Reply:
Would the whole relationship between `types` and `pendingStorageConversion` be simpler if instead you just did away with `pendingStorageConversion` entirely, had `Add` add to `f.Types`, and then had a `result astmodel.Types` which was produced by `process`?

So:
```
    inputTypes astmodel.Types // input types to be processed by process() - this is filled out by Add
```

The set of types that are `pendingStorageConversion` then is just all `inputTypes`? So you loop over that and call `createStorageVariant` for each one, except now that you're not doing it using the queue-processing function you have a bit more flexibility in return type, so you could just have `createStorageVariant` return the new `TypeDefinition` it created, which you could store in a local `result astmodel.Types` variable inside `process()`?  This code ends up looking a lot more like a mini-pipeline stage (in that it's dealing with sets of types?)

I feel like the separation of the set of input types and the set of output types helps reason about things like `createStorageVariant` more easily, as that operates on `f.types`, but if `f.types` is simultanously being modified to include the result of added storage variants there's more confusion in "what exactly is in `f.types`" than there has to be?